contracts/contracts/vault/VaultCore.sol

```diff
@@ -356,36 +358,41 @@ contract VaultCore is VaultStorage {
      * @dev Calculate difference in percent of asset allocation for a
                strategy.
      * @param _strategyAddr Address of the strategy
-     * @return int256 Difference between current and target. 18 decimals. For ex. 10%=1e17.
+     * @return unt256 Difference between current and target. 18 decimals. 
+     *  NOTE: This is relative value! not the actual percentage
      */
-    function _strategyWeightDifference(address _strategyAddr)
+    function _strategyWeightDifference(address _strategyAddr, uint256 _modAmount, bool deposit)
         internal
         view
-        returns (int256 difference)
+        returns (uint256 difference)
     {
-        difference =
-            int256(strategies[_strategyAddr].targetWeight) -
-            int256(
-                _totalValueInStrategy(_strategyAddr).divPrecisely(_totalValue())
-            );
+        // formula is (weight - valueInStrategy / totalValue)
+        // since we are comparing relative weights, we should scale by weight so
+        // that even small weights will be triggered, ie 1% versus 20%
+        // this becomes (weight - valueInStrategy / totalValue) / weight  -> (1 - valueInStrategy / (totalValue * weight))
+        // since we are comparing weight diffs in the same asset, the one and totalValue can fallout
+        // so we have (- valueInStrategy/ weight)
+        // using MAX_UINT - value to emulate the -
+        difference = MAX_UINT - (deposit ? _totalValueInStrategy(_strategyAddr).add(_modAmount) :
```

Comment:
Makes sense.

One caveat though is that if we start having strategies that support multiple assets then this formula won't apply. 

For example imagine if we have Compound for DAI and USDT. And Aave for just DAI. Compound's strategy total value will be inclusive of both DAI and USDT.

Another example is when we add the new Uniswap strategy.

So I wonder if _strategyWeightDifference and _totalValueInStrategy should take an additional parameter that is the asset address.

Reply:
We can certainly do that, is that something we want for this? the asset address is definitely there.

Reply:
Before we make more changes, I think it would be worth having a discussion with the team about where we want to go for the allocate logic...

@tomlinton wrote a proposal in #339 for doing the allocation manually which could be an alternative.

Reply:
So are we going with that? That feels like something further off in the future, while this I think will have a good immediate impact. Should reduce gas prices too.